packages/cojson-transport-ws/src/index.ts

```diff
@@ -1,108 +1,121 @@
-                    console.warn("Failed closing ws", e);
-                }
-            });
-        });
-
-        let pingTimeout: ReturnType<typeof setTimeout> | undefined;
-
-        ws.addEventListener("message", (event) => {
-            const msg = JSON.parse(event.data as string);
-
-            if (pingTimeout) {
-                clearTimeout(pingTimeout);
-            }
+        const closed = fromEventListener(ws, "close").pipe(
+            Effect.map(
+                (event) =>
+                    new DisconnectedError(`${event.code}: ${event.reason}`),
+            ),
+            Effect.flip,
+            Stream.fromEffect,
+        );
 
-            pingTimeout = setTimeout(() => {
-                console.debug("Ping timeout");
-                void Effect.runPromise(
-                    Queue.offer(incoming, Either.left(new PingTimeoutError())),
-                );
-                try {
-                    ws.close();
-                } catch (e) {
-                    console.error(
-                        "Error while trying to close ws on ping timeout",
-                        e,
-                    );
-                }
-            }, 2500);
+        const isSyncMessage = (msg: unknown): msg is SyncMessage => {
+            // boo side-effects
 
-            if (msg.type === "ping") {
-                // eslint-disable-next-line @typescript-eslint/no-explicit-any
-                (globalThis as any).jazzPings =
-                    // eslint-disable-next-line @typescript-eslint/no-explicit-any
-                    (globalThis as any).jazzPings || [];
-                // eslint-disable-next-line @typescript-eslint/no-explicit-any
-                (globalThis as any).jazzPings.push({
+            // eslint-disable-next-line @typescript-eslint/no-explicit-any
+            if ((msg as any)?.type === "ping") {
+                const ping = msg as PingMsg;
+                g.jazzPings ||= [];
+                g.jazzPings.push({
                     received: Date.now(),
-                    sent: msg.time,
-                    dc: msg.dc,
+                    sent: ping.time,
+                    dc: ping.dc,
                 });
-                return;
-            } else {
-                void Effect.runPromise(
-                    Queue.offer(incoming, Either.right(msg)),
-                );
+                return false;
             }
-        });
+            return true;
+        };
 
-        ws.addEventListener("open", () => {
-            void Stream.fromQueue(outgoing).pipe(
-                Stream.runForEach((msg) =>
-                    Effect.sync(() => ws.send(JSON.stringify(msg))),
+        yield* Queue.take(outgoing).pipe(
+            Effect.andThen((message) => ws.send(JSON.stringify(message))),
+            Effect.forever,
+            Effect.whenEffect(
+                fromEventListener(ws, "open").pipe(Effect.as(true)),
+            ),
+            Effect.forkDaemon, // :/
+        );
```

Comment:
reads as:
> `take` one entry from `outgoing` `andThen` `.send` the message. Do this `forever` `when*` the open event is emitted, which we regard `as` `true` .

I wish I would not need to `forkDaemon` here, but we can't attach to the parent scope since it's unmanaged.

Reply:
woooooah took me a while to understand this, specifically the `whenEffect` as it is changes the direction of reading in a way

What about making it a bit more explicit with a gen so "read as" comments won't be necessary?

```ts
// s/ fromEventListener/ takeFromEventListener/g

yield* Effect.gen(function*() {
  yield* fromEventListener(ws, "open");
  yield* Queue.take(outgoing).pipe(Effect.andThen(...))
});
```

Reply:
I wanted to avoid running the `fromEventListener` more than once, I could also use `Effect.once` for this but thought this was nicer. Guess not haha